File: template/ruby/commit_files.rb
#!/usr/bin/env ruby

# commit_files.rb
# to run: ruby commit_files.rb

require 'fileutils'
require 'json'
require 'digest'
require 'time'

def calculate_file_hash(file_path)
	Digest::SHA256.file(file_path).hexdigest
end

def extract_metadata(content, file_path)
	metadata = {
		'author' => '',
		'title' => File.basename(file_path),
		'hashtags' => [],
		'file_hash' => calculate_file_hash(file_path)
	}

	# Extract author
	author_match = content.match(/Author:\s*(.+)/)
	metadata['author'] = author_match[1].strip if author_match

	# Extract title (assuming it's the first line of the file)
	title_match = content.match(/^(.+)/)
	metadata['title'] = title_match[1].strip if title_match

	# Extract hashtags
	metadata['hashtags'] = content.scan(/#\w+/)

	metadata
end

def store_metadata(file_path, metadata)
	metadata_dir = File.join(File.dirname(file_path), 'metadata')
	FileUtils.mkdir_p(metadata_dir)

	metadata_file = File.join(metadata_dir, "#{File.basename(file_path)}.json")

	File.write(metadata_file, JSON.pretty_generate(metadata))

	metadata_file
end

def run_git_command(command)
	output = `#{command} 2>&1`
	[output.strip, $?.success?]
end

def commit_text_files(repo_path = ".")
	curr_dir = Dir.pwd
	repo_path = File.expand_path(repo_path)
	Dir.chdir(repo_path)

	begin
		# Get modified and untracked files
		changed_files, _ = run_git_command("git diff --name-only")
		untracked_files, _ = run_git_command("git ls-files --others --exclude-standard")

		all_files = changed_files.split("\n") + untracked_files.split("\n")
		txt_files = all_files.select { |f| f.end_with?('.txt') }

		if txt_files.empty?
			puts "No uncommitted .txt files found."
			return
		end

		# Process each file and store metadata
		metadata_files = []
		txt_files.each do |file_path|
			begin
				abs_path = File.join(repo_path, file_path)
				content = File.read(abs_path, encoding: 'UTF-8')
				metadata = extract_metadata(content, abs_path)
				metadata_file = store_metadata(abs_path, metadata)

				puts "File: #{file_path}"
				puts "Author: #{metadata['author']}"
				puts "Title: #{metadata['title']}"
				puts "Hashtags: #{metadata['hashtags'].join(', ')}"
				puts "File Hash: #{metadata['file_hash']}"
				puts

				rel_metadata = metadata_file.sub("#{repo_path}/", '')
				run_git_command("git add '#{file_path}' '#{rel_metadata}'")
				metadata_files << rel_metadata
			rescue => e
				puts "Error processing file #{file_path}: #{e.message}"
			end
		end

		# Create commit
		timestamp = Time.now.strftime('%Y-%m-%d %H:%M:%S')
		commit_message = "Auto-commit #{txt_files.size} text files and metadata on #{timestamp} by commit_files.rb"
		run_git_command("git commit -m '#{commit_message}'")

		puts "Committed #{txt_files.size} text files and their metadata."
		puts "Commit message: #{commit_message}"
	rescue => e
		puts "Error: #{e.message}"
	ensure
		Dir.chdir(curr_dir)
	end
end

if __FILE__ == $0
	commit_text_files
end

File: template/perl/commit_files.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Basename;
use File::Path qw(make_path);
use JSON;
use Digest::SHA qw(sha256_hex);
use Cwd qw(abs_path);

sub calculate_file_hash {
	my ($file_path) = @_;
	open my $fh, '<', $file_path or die "Can't open '$file_path': $!";
	my $sha = Digest::SHA->new(256);
	$sha->addfile($fh);
	close $fh;
	return $sha->hexdigest;
}

sub extract_metadata {
	my ($content, $file_path) = @_;
	my %metadata = (
		author => '',
		title => basename($file_path),
		hashtags => [],
		file_hash => calculate_file_hash($file_path)
	);

	if ($content =~ /Author:\s*(.+)/m) {
		$metadata{author} = $1;
	}

	if ($content =~ /^(.+)/m) {
		$metadata{title} = $1;
	}

	$metadata{hashtags} = [$content =~ /#(\w+)/g];

	return \%metadata;
}

sub store_metadata {
	my ($file_path, $metadata) = @_;
	my $metadata_dir = dirname($file_path) . '/metadata';
	make_path($metadata_dir);

	my $metadata_file = $metadata_dir . '/' . basename($file_path) . '.json';
	open my $fh, '>', $metadata_file or die "Can't open '$metadata_file': $!";
	print $fh JSON->new->pretty->encode($metadata);
	close $fh;

	return $metadata_file;
}

sub run_git_command {
	my ($command) = @_;
	my $output = `$command 2>&1`;
	return $output;
}

sub commit_text_files {
	my $repo_path = shift || '.';
	my $curr_dir = `pwd`;
	chomp $curr_dir;
	$repo_path = Cwd::abs_path($repo_path);
	chdir $repo_path;

	eval {
		# Get modified and untracked files
		my $changed_files = run_git_command("git diff --name-only");
		my $untracked_files = run_git_command("git ls-files --others --exclude-standard");

		my @all_files = split(/\n/, $changed_files . $untracked_files);
		my @txt_files = grep { /\.txt$/ } @all_files;

		if (!@txt_files) {
			print "No uncommitted .txt files found.\n";
			return;
		}

		foreach my $file_path (@txt_files) {
			eval {
				my $abs_path = "$repo_path/$file_path";
				open my $fh, '<:encoding(UTF-8)', $abs_path or die "Can't open '$abs_path': $!";
				my $content = do { local $/; <$fh> };
				close $fh;

				my $metadata = extract_metadata($content, $abs_path);
				my $metadata_file = store_metadata($abs_path, $metadata);

				print "File: $file_path\n";
				print "Author: $metadata->{author}\n";
				print "Title: $metadata->{title}\n";
				print "Hashtags: " . join(', ', @{$metadata->{hashtags}}) . "\n";
				print "File Hash: $metadata->{file_hash}\n\n";

				my $rel_metadata = $metadata_file;
				$rel_metadata =~ s/^\Q$repo_path\E\///;
				run_git_command("git add '$file_path' '$rel_metadata'");
			};
			if ($@) {
				print "Error processing file $file_path: $@\n";
			}
		}

		# Create commit
		my $timestamp = `date '+%Y-%m-%d %H:%M:%S'`;
		chomp $timestamp;
		my $commit_message = "Auto-commit " . scalar(@txt_files) . " text files and metadata on $timestamp by commit_files.pl";
		run_git_command("git commit -m '$commit_message'");

		print "Committed " . scalar(@txt_files) . " text files and their metadata.\n";
		print "Commit message: $commit_message\n";
	};
	if ($@) {
		print "Error: $@\n";
	}

	chdir $curr_dir;
}

commit_text_files();